### PR TITLE
Group reader

### PR DESCRIPTION
**Pull request by poy**
The GroupReader allows a consumer to read from several source IDs.

### CODE REVIEW

**Comment by poy** [comment]
Related PRs:

https://github.com/cloudfoundry-incubator/go-log-cache/pull/2
https://github.com/cloudfoundry-incubator/log-cache-release/pull/5

File: api/group_reader.proto [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    rpc RemoveFromGroup(RemoveFromGroupRequest) returns (RemoveFromGroupResponse) {
+        option (google.api.http) = {
+            delete: "/v1/group/{name}/{source_id}"
+        };
+    }
+
+    rpc Read(GroupReadRequest) returns (GroupReadResponse) {
+        option (google.api.http) = {
+            get: "/v1/group/read/{name}"
```
Comment (bradylove) [inline api/group_reader.proto]: I feel like this endpoint would better fit into REST if it were `/v1/group/{name}/envelopes`. `GET` to me, is inherently a `READ` action.

File: internal/groups/manager.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		m: make(map[string][]string),
+		s: s,
+	}
+}
+
+// AddToGroup creates the given group if it does not exist or adds the
+// sourceID if it does.
+func (m *Manager) AddToGroup(ctx context.Context, r *logcache.AddToGroupRequest, _ ...grpc.CallOption) (*logcache.AddToGroupResponse, error) {
+	if r.GetName() == "" || r.GetSourceId() == "" {
+		return nil, errors.New("name and source_id fields are required")
```
Comment (bradylove) [inline internal/groups/manager.go]: We should use the gRPC error codes here and anywhere else we return errors.

**Review by bradylove — COMMENTED** [review]
Some concerns I have with this are:

* Two clients choosing the same name for their group and stepping on each others toes.
* No automatic cleanup of groups when clients connect/disconnect.
* No size limit to the name. I can connect and create as many groups as I want with names as large as I want (as long as they are under the gRPC MaxMsgSize)

These could be resolved by assigning the client an ID for the group rather than the client specifying a name when creating a new group.

**Comment by poy** [comment]
1. I want the clients to be able to choose their group names so they can easily coordinate.I think later on authorization *should* prevent two different users from creating and accessing the same group.
1. I'm not sure how groups should clean up. Perhaps they expire? This seems like something that can easily be added later.
1. I updated the API to:

```
patch: "/v1/group/{name}/{source_id}"
delete: "/v1/group/{name}/{source_id}"
get: "/v1/group/{name}"
get: "/v1/group/{name}/meta"
```

**Pull request merged**